### PR TITLE
Support both train / eval modes for ModuleInfo

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78735

This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It makes the following changes:
* Adds a required `training` arg to `ModuleInfo.module_inputs_func` - this allows those functions to generate sample module inputs based on the training mode.
* Updates the `@modules` decorator to additionally pass a `training` arg to test functions.
    * The `training` arg values to pass can be customized via `TrainEvalMode` e.g. `@modules(..., train_eval_mode=TrainEvalMode.train_only)`. Supported modes are `train_only`, `eval_only`, and `train_and_eval`.
    * Tests that use the decorator in `test/test_modules.py` have been updated. They now ingest the `training` arg, pass it to the `module_inputs_func`, and set the training mode for instantiated module instances with `m.train(training)`. There's almost certainly a way to do this with less boilerplate.
* Adds a new `train_and_eval_differ` option to `ModuleInfo` for indicating whether a specific module cares about the training mode.
    * If this is `False`, tests are generated just for a single mode. This avoids pointlessly running tests twice when the training mode setting doesn't matter.
    * Adds a new test in `test/test_modules.py` to verify that `train_and_eval_differ` is set correctly. It does this by instantiating each module, deleting the `training` attribute, and then running the forward pass. If the forward pass references `training`, it will throw an `AttributeError`, indicating that `train_and_eval_differ=True` should be set.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
SGTM!
Just minor comments

File: test/test_modules.py [hunk 1]
```diff
@@ -233,15 +241,15 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @skipMeta
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_modules.py:244]: Just reordered it slightly for consistency, before this was underneath `@modules`

File: test/test_modules.py [hunk 2]
```diff
@@ -640,6 +654,43 @@ def inner_check_out_mem_format(output):
...
+                    m(*args, **kwargs)
+                except AttributeError as e:
+                    if "'training'" in str(e):
+                        self.assertTrue(module_info.train_and_eval_differ,
+                                        f"The ModuleInfo entry for {module_info.name} has "
+                                        "train_and_eval_differ=False, but the training mode was found to "
+                                        "affect the forward pass. Consider setting train_and_eval_differ=True "
+                                        "for this ModuleInfo entry.")
+                    else:
+                        raise
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_modules.py:693]: That's my understanding, but I'll change it to `raise e` for clarity

File: test/test_modules.py [hunk 3]
```diff
@@ -640,6 +654,43 @@ def inner_check_out_mem_format(output):
...
+        module_cls = module_info.module_cls
+        module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
+                                                       requires_grad=False, training=training)
+
+        # Run forward inputs through to see if the training flag is accessed during forward.
+        for module_input in module_inputs:
+            if module_input.forward_input is None:
+                continue
+
+            with freeze_rng_state():
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_modules.py:671]: Not strictly needed here; this is used in other tests for determinism. Removing

**Review by jbschlosser — COMMENTED** [review]
@albanD Before I can merge this, it looks like I need a way to skip a test for eval mode only.

I have a working fix that can skip based on anything passed to the test - going to add a commit earlier in the PR stack that adds this

File: test/test_modules.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_modules.py:244]: Ho ok

**Review by albanD — APPROVED** [review]
